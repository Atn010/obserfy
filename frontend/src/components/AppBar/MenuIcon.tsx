import React, { FC, MouseEventHandler } from "react"
import Icon from "../Icon/Icon"
import { ReactComponent as Menu } from "../../icons/menu-outline.svg"
import { Box } from "../Box/Box"

interface Props {
  onMenuClick?: MouseEventHandler<HTMLImageElement>
}
const MenuIcon: FC<Props> = ({ onMenuClick }) => (
  <Box
    onClick={onMenuClick}
    maxWidth={56}
    sx={{
      cursor: "pointer",
      pointerEvents: ["inherit", "none"],
    }}
  >
    <Icon
      as={Menu}
      data-cy="menu"
      alt="Menu"
      width={["icon", 0]}
      ml={[3, 100]} // 100 is for sliding effect when viewport resize from md to sm
      opacity={[1, 0]}
      size={24}
      // TODO: This animation is everywhere, consider creating animation framework
      sx={{ transition: "margin-left 250ms cubic-bezier(0.0, 0.0, 0.2, 1)" }}
    />
  </Box>
)
export default MenuIcon
