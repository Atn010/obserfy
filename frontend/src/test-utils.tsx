/* eslint-disable import/no-extraneous-dependencies */
import React, { FC, ReactElement } from "react"
import { ThemeProvider } from "theme-ui"
import { render, RenderOptions, RenderResult } from "@testing-library/react"
import { IntlProvider } from "gatsby-plugin-intl3"
import {
  createHistory,
  createMemorySource,
  LocationProvider,
} from "@reach/router"
import Theme from "./gatsby-plugin-theme-ui"
import Box from "./components/Box/Box"
import en from "./intl/en.json"

const AllProviders: FC = ({ children }) => {
  // TODO: Maybe this better be outside
  const history = createHistory(createMemorySource("/"))
  return (
    <LocationProvider history={history}>
      <ThemeProvider theme={Theme}>
        <IntlProvider locale="en" messages={en}>
          <Box fontSize={[16, 20]}>{children}</Box>
        </IntlProvider>
      </ThemeProvider>
    </LocationProvider>
  )
}

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "queries">
): RenderResult => render(ui, { wrapper: AllProviders, ...options })

// re-export everything
export * from "@testing-library/react"

// override render method
export { customRender as render }
