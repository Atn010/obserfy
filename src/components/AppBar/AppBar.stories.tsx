import React, { FC } from "react"
import AppBar from "./AppBar"

export default {
  title: "Complex|AppBar",
  component: AppBar,
  parameters: {
    componentSubtitle: "Just a simple AppBar"
  }
}

export const Basic: FC = () => {
  return <AppBar title="Home" />
}
