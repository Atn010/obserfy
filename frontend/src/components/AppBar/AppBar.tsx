import React, {
  FC,
  MouseEventHandler,
  useEffect,
  useRef,
  useState,
} from "react"
import GatsbyImage from "gatsby-image"
import { navigate } from "gatsby"
import { Link } from "gatsby-plugin-intl3"
import Typography from "../Typography/Typography"
import { ReactComponent as BookIcon } from "../../icons/book.svg"
import { ReactComponent as LogoutIcon } from "../../icons/logout.svg"
import Avatar from "../Avatar/Avatar"
import Icon from "../Icon/Icon"
import { ReactComponent as ArrowDownIcon } from "../../icons/arrow-down-outline.svg"
import { Card } from "../Card/Card"
import { useOutsideClick } from "../../hooks"
import Spacer from "../Spacer/Spacer"
import MenuIcon from "./MenuIcon"
import Flex, { FlexProps } from "../Flex/Flex"
import Box from "../Box/Box"
import { useAvatarPlaceholder } from "../../useAvatarPlaceholder"
import useOldApiHook from "../../api/useOldApiHook"
import { getSchoolId } from "../../hooks/schoolIdState"
import { getAnalytics } from "../../analytics"
import {
  setCrispCompanyName,
  setCrispEmail,
  setCrispNickName,
  setCrispToken,
} from "../../crisp"
import Button from "../Button/Button"

interface Props {
  onMenuClick?: MouseEventHandler<HTMLImageElement>
  position?: "fixed" | "relative"
  title: string
}

export const AppBar: FC<Props> = ({
  onMenuClick,
  title,
  position = "relative",
}) => {
  return (
    <Card
      as="header"
      backgroundColor="surface"
      height="appbar"
      width="100%"
      display="flex"
      borderRadius={0}
      sx={{
        alignItems: "center",
        borderColor: "border",
        zIndex: 50,
        top: 0,
        position,
      }}
    >
      <SchoolName
        alignItems="center"
        height={57}
        width={[0, "auto"]}
        overflow="hidden"
      />
      <MenuIcon onMenuClick={onMenuClick} />
      <Typography.H6 fontWeight="body" opacity={[1, 0]} mb={0} ml={[0, 3]}>
        {title}
      </Typography.H6>
      <Spacer />
      <UserAvatar />
    </Card>
  )
}

export const SchoolName: FC<FlexProps> = ({ ...props }) => {
  const schoolId = getSchoolId()
  const [school] = useOldApiHook<{ name: string }>(`/schools/${schoolId}`)

  useEffect(() => {
    if (school) {
      getAnalytics()?.identify({
        schoolName: school.name,
      })
      setCrispCompanyName(school.name)
    }
  }, [school])

  return (
    <Flex alignItems="center" height="appbar" {...props}>
      <Link to="/choose-school">
        <Button variant="secondary">
          <Icon
            as={BookIcon}
            minWidth={24}
            size={24}
            m={0}
            mr={3}
            alt="School Icon"
          />
          <Typography.Body
            as="div"
            ml="-4px"
            mb={0}
            sx={{ whiteSpace: "nowrap" }}
          >
            {school?.name}
          </Typography.Body>
        </Button>
      </Link>
    </Flex>
  )
}

const UserAvatar: FC = () => {
  const [isShowingOption, setIsShowingOption] = useState(false)
  const element = useRef<HTMLElement>(null)
  const avatar = useAvatarPlaceholder()
  useOutsideClick(element, () => setIsShowingOption(false))
  const [userData] = useOldApiHook<{
    id: string
    name: string
    email: string
  }>("/user")

  useEffect(() => {
    if (userData) {
      getAnalytics()?.identify(userData.id, {
        name: userData.name,
        email: userData.email,
      })
      setCrispToken(userData.id)
      setCrispEmail(userData.email)
      setCrispNickName(userData.name.split(" ")[0])
    }
  }, [userData])

  return (
    <Flex
      ref={element}
      mx={3}
      p={1}
      alignItems="center"
      onClick={() => setIsShowingOption(!isShowingOption)}
      sx={{
        borderStyle: "solid",
        borderRadius: "circle",
        borderColor: "border",
        borderWidth: 1,
        cursor: "pointer",
        transition: "background-color 250ms cubic-bezier(0.0, 0.0, 0.2, 1)",
        "&:hover": {
          backgroundColor: "muted",
        },
      }}
      data-cy="schoolName"
    >
      <Avatar as={Box}>
        <GatsbyImage fixed={avatar} alt="avatar" />
      </Avatar>
      <Typography.Body
        as="div"
        ml={2}
        mr={1}
        my={0}
        display={["none", "block"]}
      >
        {userData?.name}
      </Typography.Body>
      <Icon
        as={ArrowDownIcon}
        alt="Store Icon"
        m={0}
        ml={[2, 0]}
        mr={[2, 1]}
        opacity={0.4}
      />
      {isShowingOption && <SignOutCard />}
    </Flex>
  )
}

const SignOutCard: FC = () => {
  async function logout(): Promise<void> {
    const response = await fetch("/auth/logout", {
      method: "POST",
      credentials: "same-origin",
    })
    if (response.status === 200) {
      navigate("/login")
    }
  }

  return (
    <Card
      onClick={logout}
      sx={{
        top: 52,
        right: 3,
        position: "fixed",
        boxShadow: "low",
        borderRadius: "default",
      }}
    >
      <Flex alignItems="center" width={140}>
        <Icon as={LogoutIcon} size={20} />
        <Typography.Body m={0}>Log out</Typography.Body>
      </Flex>
    </Card>
  )
}

export default AppBar
