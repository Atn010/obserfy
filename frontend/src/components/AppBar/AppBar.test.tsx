import React from "react"
import { render } from "../../test-utils"
import { Basic } from "./AppBar.stories"

describe("AppBar", () => {
  it("should render correctly", () => {
    const { container } = render(<Basic />)
    expect(container).toMatchSnapshot()
  })
})
