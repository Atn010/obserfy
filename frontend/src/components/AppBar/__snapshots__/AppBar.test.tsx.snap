// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppBar should render correctly 1`] = `
.emotion-12 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  font-size: 16px;
}

@media screen and (min-width:40em) {
  .emotion-12 {
    font-size: 20px;
  }
}

.emotion-11 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  padding: 8px;
  background-color: var(--theme-ui-colors-background,#F7F8FA);
  border-radius: 6px;
  box-shadow: rgba(0,0,0,0.1) 0px 0px 1px,rgba(0,0,0,0.27) 0px 2px 4px -2px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-color: var(--theme-ui-colors-border,rgba(0,0,0,.1));
  z-index: 50;
  top: 0;
  position: relative;
  border-radius: 0;
  padding: 0;
  background-color: var(--theme-ui-colors-surface,#fff);
  overflow: hidden;
  height: 57px;
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-2 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  height: 57px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 0%;
  overflow: hidden;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

@media screen and (min-width:40em) {
  .emotion-2 {
    width: auto;
  }
}

.emotion-0 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  max-width: 100%;
  height: auto;
  fill: var(--theme-ui-colors-icon,rgba(0,0,0,0.6));
  margin: 16px;
  height: 24px;
  width: 24px;
}

.emotion-1 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  color: rgba(0,0,0,0.88);
  font-size: 1.6rem;
  font-family: Open Sans,-apple-system,BlinkMacSystemFont,San Francisco,Roboto,Segoe UI,Helvetica Neue,sans-serif;
  font-weight: 400;
  line-height: 3rem;
  white-space: nowrap;
  color: var(--theme-ui-colors-text,rgba(0,0,0,0.87));
  font-size: 16px;
  margin-left: -4px;
  margin-bottom: 0;
}

@media screen and (min-width:40em) {
  .emotion-1 {
    font-size: 14px;
  }
}

.emotion-4 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  cursor: pointer;
  pointer-events: inherit;
  max-width: 56px;
}

@media screen and (min-width:40em) {
  .emotion-4 {
    pointer-events: none;
  }
}

.emotion-3 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  max-width: 100%;
  height: auto;
  -webkit-transition: margin-left 250ms cubic-bezier(0.0,0.0,0.2,1);
  transition: margin-left 250ms cubic-bezier(0.0,0.0,0.2,1);
  fill: var(--theme-ui-colors-icon,rgba(0,0,0,0.6));
  margin: 16px;
  height: 24px;
  width: 24px;
  margin-left: 16px;
  opacity: 1;
}

@media screen and (min-width:40em) {
  .emotion-3 {
    width: 0%;
    margin-left: 100px;
    opacity: 0;
  }
}

.emotion-5 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  font-size: 24px;
  font-family: Open Sans,-apple-system,BlinkMacSystemFont,San Francisco,Roboto,Segoe UI,Helvetica Neue,sans-serif;
  font-weight: 300;
  line-height: heading;
  font-size: 2rem;
  line-height: 3rem;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  color: var(--theme-ui-colors-text,rgba(0,0,0,0.87));
  font-weight: 400;
  opacity: 1;
  margin-bottom: 0;
  margin-left: 0;
}

@media screen and (min-width:40em) {
  .emotion-5 {
    opacity: 0;
    margin-left: 16px;
  }
}

.emotion-6 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  pointer-events: none;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.emotion-10 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  border-style: solid;
  border-radius: 99999px;
  border-color: var(--theme-ui-colors-border,rgba(0,0,0,.1));
  border-width: 1px;
  cursor: pointer;
  -webkit-transition: background-color 250ms cubic-bezier(0.0,0.0,0.2,1);
  transition: background-color 250ms cubic-bezier(0.0,0.0,0.2,1);
  margin-left: 16px;
  margin-right: 16px;
  padding: 4px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-10:hover {
  background-color: var(--theme-ui-colors-muted,#f6f6f9);
}

.emotion-7 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  max-width: 100%;
  height: auto;
  object-fit: cover;
  object-position: center;
  border-radius: 99999px;
  border-color: var(--theme-ui-colors-border,rgba(0,0,0,.1));
  border-width: 1px;
  border-style: solid;
  overflow: hidden;
  width: 32px;
  height: 32px;
}

.emotion-8 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  color: rgba(0,0,0,0.88);
  font-size: 1.6rem;
  font-family: Open Sans,-apple-system,BlinkMacSystemFont,San Francisco,Roboto,Segoe UI,Helvetica Neue,sans-serif;
  font-weight: 400;
  line-height: 3rem;
  color: var(--theme-ui-colors-text,rgba(0,0,0,0.87));
  font-size: 16px;
  margin-left: 8px;
  margin-right: 4px;
  margin-top: 0;
  margin-bottom: 0;
  display: none;
}

@media screen and (min-width:40em) {
  .emotion-8 {
    font-size: 14px;
    display: block;
  }
}

.emotion-9 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  max-width: 100%;
  height: auto;
  fill: var(--theme-ui-colors-icon,rgba(0,0,0,0.6));
  margin: 0;
  height: 20px;
  width: 20px;
  margin-left: 8px;
  margin-right: 8px;
  opacity: 0.4;
}

@media screen and (min-width:40em) {
  .emotion-9 {
    margin-left: 0;
    margin-right: 4px;
  }
}

<div>
  <div
    class="emotion-12"
  >
    <header
      class="emotion-11"
    >
      <div
        class="emotion-2"
      >
        <div
          alt="School Icon"
          class="emotion-0"
        />
        <div
          class="emotion-1"
        />
      </div>
      <div
        class="emotion-4"
      >
        <div
          alt="Menu"
          class="emotion-3"
          data-cy="menu"
        />
      </div>
      <h6
        class="emotion-5"
      >
        Home
      </h6>
      <div
        class="emotion-6"
      />
      <div
        class="emotion-10"
        data-cy="schoolName"
      >
        <div
          class="emotion-7"
        >
          <div
            class=" gatsby-image-wrapper"
            style="position: relative; overflow: hidden; display: inline-block; width: 32px; height: 32px;"
          >
            <img
              alt=""
              src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABQAAAAUCAYAAACNiR0NAAAACXBIWXMAAAsTAAALEwEAmpwYAAADvUlEQVQ4y7WU7VNUZRjG98/oSzON9ikTQUFidyHY1MDBlCgBqxmbUacGh3Eop9LsxSY0NN6FaUEQWCxCZYBKBxVLS1wQAiZCQNBVCY2MJHZh2XOeX/fZl0Bt+lRn5tpznjnn/s113c/9rIn/+DL9r0CllF/zz7pI898JrXUNtAXSA++CRfPA+0BG0QOX+jdbBtRQyOGDrkII3+Qg3juX0WcmYO4e2swg+q12ON+G+u40qusS3Bm/D2wKFS+MO3vjG+62pvJ78wqmWyPwOZNQPcnwYxxk21BrY2G9FTbIOiMJct+DuxNBhwbIQAZjugermfhyMZ6zEdAhhV1S5DRDbwwtb4RzJV6e09eg0gLixVWQGAVZm8HjDkUOwHTvNPda49AuPgXdCajOWHSnQPvjOFMQRYZ1OcUvPwuvPodKfQa1cTUY2rRWoCvhcLHhUPu74cpVinJGw+WnUR3WAKzXwrelYXz97lZ8J/LwtBThfUvcbIgXh0GgoecTYFtasIeaBzW8Gy4tB3FlwDQD1mPml+ZV1L/5Er7GQjoP7GDAvoee11IhJcHvUBmR08X1OgsU7xegx4Xq2wTt4QIznAmoQ/rWa0Y/n4a74SB/1h+U+wGGy3bx/afZDGxLgVTbfOSNwV7ecgmwO1GcRUpMieCMDQD7Y2nKXUrmmmiuHNrth/3xRS6TR/dxsWAn7uxX4AXpYYbUpotWS7KifYFdVk6rNF/siyu9R0ah28pErpWUJYswP/4IF3IyobmQqbqP+c2Rg6t4pzizQVI0KjkGZUQ1xmZ2Jjg2nYY7I24keqNEzUridqKNjGWPkhS+mLHDe+kqeIeJJjvTTSXMtR1HNTjQK0rQa+wyn53zJ8kYbO+1U7h/bmC2Tz5sr4eb5zhW8D5xix4jf6sNz4m9jDp2MTU2JJvnffjU+WdYGT9+qOm49Pdz6WmDHIRTW65ztaKMuS4bA2VZDH3UyrW8o4zub6Rt8whfbYGmzF859rqLszm3Ge9zP3TKTEdWaFRH+aiOVFRKcvsT8MP2kzAQwYWsFuxPQqV0pVL6XrVMURHmIz/8JrlhQ+RHjeCsmgw41TQ/1FRrVtTGIFI4LBoOq4/PBHq9XGKeTOOInCqHWfer1qKokz2otngpiRulyDLCJ0uG+allKvTvJUA/LACskRPnsPioWCqbvcfBzLlk6qxz1EQH3wdVK98dMo9RFHuVvJUuqtf3Mtl5Bs09tRAYcumjXIA9H1Qx07bOv/4nYKl5nELLMIVmF2XxffR/uJ0b5W/zF4A0F9hGaJU9AAAAAElFTkSuQmCC"
              style="position: absolute; top: 0px; left: 0px; width: 100%; height: 100%; object-fit: cover; object-position: center; opacity: 1; transition-delay: 500ms;"
            />
            <picture>
              <source
                srcset="/static/d765df87b7164d5d26f62ef45d04c06f/62cc5/avatar.webp 1x,
/static/d765df87b7164d5d26f62ef45d04c06f/0b7e8/avatar.webp 1.5x,
/static/d765df87b7164d5d26f62ef45d04c06f/adbdf/avatar.webp 2x"
                type="image/webp"
              />
              <source
                srcset="/static/d765df87b7164d5d26f62ef45d04c06f/fa1da/avatar.png 1x,
/static/d765df87b7164d5d26f62ef45d04c06f/984fe/avatar.png 1.5x,
/static/d765df87b7164d5d26f62ef45d04c06f/b067b/avatar.png 2x"
              />
              <img
                alt="avatar"
                height="32"
                loading="lazy"
                src="/static/d765df87b7164d5d26f62ef45d04c06f/fa1da/avatar.png"
                srcset="/static/d765df87b7164d5d26f62ef45d04c06f/fa1da/avatar.png 1x,
/static/d765df87b7164d5d26f62ef45d04c06f/984fe/avatar.png 1.5x,
/static/d765df87b7164d5d26f62ef45d04c06f/b067b/avatar.png 2x"
                style="position: absolute; top: 0px; left: 0px; width: 100%; height: 100%; object-fit: cover; object-position: center; opacity: 0; transition: opacity 500ms;"
                width="32"
              />
            </picture>
            <noscript>
              &lt;picture&gt;&lt;source type='image/webp' srcset="/static/d765df87b7164d5d26f62ef45d04c06f/62cc5/avatar.webp 1x,
/static/d765df87b7164d5d26f62ef45d04c06f/0b7e8/avatar.webp 1.5x,
/static/d765df87b7164d5d26f62ef45d04c06f/adbdf/avatar.webp 2x" /&gt;&lt;source srcset="/static/d765df87b7164d5d26f62ef45d04c06f/fa1da/avatar.png 1x,
/static/d765df87b7164d5d26f62ef45d04c06f/984fe/avatar.png 1.5x,
/static/d765df87b7164d5d26f62ef45d04c06f/b067b/avatar.png 2x" /&gt;&lt;img loading="lazy" width="32" height="32" srcset="/static/d765df87b7164d5d26f62ef45d04c06f/fa1da/avatar.png 1x,
/static/d765df87b7164d5d26f62ef45d04c06f/984fe/avatar.png 1.5x,
/static/d765df87b7164d5d26f62ef45d04c06f/b067b/avatar.png 2x" src="/static/d765df87b7164d5d26f62ef45d04c06f/fa1da/avatar.png" alt="avatar" style="position:absolute;top:0;left:0;opacity:1;width:100%;height:100%;object-fit:cover;object-position:center"/&gt;&lt;/picture&gt;
            </noscript>
          </div>
        </div>
        <div
          class="emotion-8"
        />
        <div
          alt="Store Icon"
          class="emotion-9"
        />
      </div>
    </header>
  </div>
</div>
`;
